File: mobile/packages/divine_video_player/ios/Classes/DivineVideoPlayerPlugin.swift
import AVFoundation
import Flutter
import UIKit

/// Entry point for the divine_video_player plugin on iOS.
///
/// Manages the lifecycle of ``DivineVideoPlayerInstance`` objects and
/// registers the platform view factory for rendering.
public class DivineVideoPlayerPlugin: NSObject, FlutterPlugin {

    private static var registrar: FlutterPluginRegistrar?

    public static func register(with registrar: FlutterPluginRegistrar) {
        // Hot restart re-calls register(with:) without disposing the
        // previous engine's players. Clean up zombie players so timers
        // and observers are released.
        PlayerRegistry.shared.disposeAll()

        self.registrar = registrar

        let globalChannel = FlutterMethodChannel(
            name: "divine_video_player",
            binaryMessenger: registrar.messenger()
        )
        let plugin = DivineVideoPlayerPlugin()
        registrar.addMethodCallDelegate(plugin, channel: globalChannel)

        registrar.register(
            DivineVideoPlayerViewFactory(messenger: registrar.messenger()),
            withId: "divine_video_player_view"
        )

        // Observe app lifecycle to pause/resume all players.
        NotificationCenter.default.addObserver(
            plugin,
            selector: #selector(appWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            plugin,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
    }

    @objc private func appWillResignActive() {
        PlayerRegistry.shared.forAll { $0.onAppBackgrounded() }
    }

    @objc private func appDidBecomeActive() {
        PlayerRegistry.shared.forAll { $0.onAppForegrounded() }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // Methods that require no arguments are handled before the
        // args guard to avoid returning FlutterMethodNotImplemented.
        if call.method == "disposeAll" {
            PlayerRegistry.shared.disposeAll()
            result(nil)
            return
        }

        guard let args = call.arguments as? [String: Any] else {
            result(FlutterMethodNotImplemented)
            return
        }

        switch call.method {
        case "create":
            guard let id = args["id"] as? Int,
                  let registrar = Self.registrar else {
                result(
                    FlutterError(
                        code: "INVALID_ARGS",
                        message: "Missing player id",
                        details: nil
                    )
                )
                return
            }
            // Dispose any existing player with the same ID before
            // creating the new one to avoid leaking zombie players.
            PlayerRegistry.shared.remove(id)?.dispose()

            let instance = DivineVideoPlayerInstance(
                messenger: registrar.messenger(),
                playerId: id
            )
            PlayerRegistry.shared.set(instance, for: id)

            let useTexture = args["useTexture"] as? Bool ?? false
            if useTexture {
                let textureId = instance.enableTextureOutput(
                    registry: registrar.textures()
                )
                result(["textureId": textureId])
            } else {
                result(nil)
            }

        case "dispose":
            guard let id = args["id"] as? Int else {
                result(nil)
                return
            }
            PlayerRegistry.shared.remove(id)?.dispose()
            result(nil)

        case "preload":
            let clips = args["clips"] as? [[String: Any]] ?? []
            Self.handlePreload(clips: clips, result: result)

        case "configureCache":
            let maxSizeBytes = args["maxSizeBytes"] as? Int ?? (500 * 1024 * 1024)
            // 10% of disk budget for in-memory cache, rest on disk.
            let memoryCapacity = maxSizeBytes / 10
            URLCache.shared = URLCache(
                memoryCapacity: memoryCapacity,
                diskCapacity: maxSizeBytes,
                diskPath: "divine_video_cache"
            )
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /// Preloads video metadata and initial buffer data by loading
    /// `AVURLAsset` properties asynchronously. The OS-level URL cache
    /// retains the fetched data so that a real player starts faster.
    private static func handlePreload(
        clips: [[String: Any]],
        result: @escaping FlutterResult
    ) {
        guard !clips.isEmpty else {
            result(nil)
            return
        }

        let group = DispatchGroup()

        for clipMap in clips {
            guard let uri = clipMap["uri"] as? String else { continue }

            let url: URL
            if uri.hasPrefix("/") {
                url = URL(fileURLWithPath: uri)
            } else if let parsed = URL(string: uri) {
                url = parsed
            } else {
                continue
            }

            group.enter()
            let asset = AVURLAsset(url: url)
            asset.loadValuesAsynchronously(forKeys: ["duration", "tracks"]) {
                group.leave()
            }
        }

        group.notify(queue: .main) {
            result(nil)
        }
    }
}

/// Global registry so that ``DivineVideoPlayerViewFactory`` can find
/// instances created during the `create` method call.
final class PlayerRegistry {
    static let shared = PlayerRegistry()
    private var players: [Int: DivineVideoPlayerInstance] = [:]
    private init() {}

    func get(_ id: Int) -> DivineVideoPlayerInstance? { players[id] }
    func set(_ instance: DivineVideoPlayerInstance, for id: Int) { players[id] = instance }
    @discardableResult
    func remove(_ id: Int) -> DivineVideoPlayerInstance? {
        let instance = players.removeValue(forKey: id)
        return instance
    }
    func disposeAll() {
        players.values.forEach { $0.dispose() }
        players.removeAll()
    }
    func forAll(_ action: (DivineVideoPlayerInstance) -> Void) {
        players.values.forEach(action)
    }
}

File: mobile/packages/divine_video_player/android/src/main/kotlin/com/divinevideo/divine_video_player/DivineVideoPlayerPlugin.kt
package com.divinevideo.divine_video_player

import android.app.Activity
import android.content.Context
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.source.DefaultMediaSourceFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.ConcurrentHashMap

/**
 * Entry point for the divine_video_player plugin on Android.
 *
 * Manages the lifecycle of [DivineVideoPlayerInstance] objects and
 * registers the platform view factory for rendering.
 */
class DivineVideoPlayerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {

    private lateinit var globalChannel: MethodChannel
    private lateinit var binding: FlutterPlugin.FlutterPluginBinding

    companion object {
        // Tracks the plugin instance attached to the main FlutterEngine.
        // Flutter creates a NEW DivineVideoPlayerPlugin instance per engine,
        // so instance variables cannot distinguish engines — only a static
        // reference to the known-main instance works.
        //
        // Background engines (Firebase FCM isolate, WorkManager) attach a
        // different instance while mainPluginInstance is non-null. They are
        // skipped entirely so PlayerRegistry.disposeAll() never kills live
        // players owned by the main engine.
        @Volatile
        private var mainPluginInstance: DivineVideoPlayerPlugin? = null
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        // Treat as the main engine when:
        //   • nothing is registered yet (cold start), OR
        //   • this same instance is reattaching (hot restart without a
        //     preceding onDetachedFromEngine call).
        // Any other instance attaching while the main one is registered is
        // a background engine — skip it.
        val isMainEngine = mainPluginInstance == null || mainPluginInstance === this
        if (!isMainEngine) return

        mainPluginInstance = this
        binding = flutterPluginBinding

        // Hot restart re-calls onAttachedToEngine on the same instance without
        // a preceding onDetachedFromEngine. Dispose zombie players left over
        // from the previous Dart VM. main.dart also calls disposeAll() before
        // runApp() as belt-and-suspenders coverage.
        PlayerRegistry.disposeAll()

        globalChannel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "divine_video_player",
        )
        globalChannel.setMethodCallHandler(this)

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "divine_video_player_view",
            DivineVideoPlayerViewFactory(flutterPluginBinding),
        )
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        if (mainPluginInstance !== this) {
            // Background engine detaching — do not touch main engine's players.
            return
        }
        mainPluginInstance = null
        globalChannel.setMethodCallHandler(null)
        PlayerRegistry.disposeAll()
        VideoCache.release()
    }

    // -- ActivityAware: pause/resume players with app lifecycle --

    private val lifecycleObserver = object : DefaultLifecycleObserver {
        override fun onPause(owner: LifecycleOwner) {
            PlayerRegistry.forAll { it.onAppBackgrounded() }
        }

        override fun onResume(owner: LifecycleOwner) {
            PlayerRegistry.forAll { it.onAppForegrounded() }
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        (binding.activity as? LifecycleOwner)?.lifecycle?.addObserver(lifecycleObserver)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        // No-op: observer is removed when activity is destroyed.
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        (binding.activity as? LifecycleOwner)?.lifecycle?.addObserver(lifecycleObserver)
    }

    override fun onDetachedFromActivity() {
        // Stop every live player before the Activity unwinds; full release
        // happens later on engine detach. See #3416.
        PlayerRegistry.forAll { it.stopForActivityDetach() }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "create" -> {
                val id = call.argument<Int>("id")!!
                // Dispose any existing player with the same ID BEFORE
                // creating the new one. The new instance registers
                // MethodChannel/EventChannel handlers under the same
                // channel name, so the old instance must release them
                // first to avoid nullifying the new handlers.
                PlayerRegistry.remove(id)?.dispose()
                val instance = DivineVideoPlayerInstance(
                    binding.binaryMessenger,
                    binding.applicationContext,
                    id,
                )
                PlayerRegistry.put(id, instance)

                val useTexture = call.argument<Boolean>("useTexture") ?: false
                if (useTexture) {
                    val useLegacySurface =
                        call.argument<Boolean>("useLegacySurface") ?: true
                    val textureId = instance.enableTextureOutput(
                        binding.textureRegistry,
                        useLegacySurface = useLegacySurface,
                    )
                    result.success(mapOf("textureId" to textureId))
                } else {
                    result.success(null)
                }
            }
            "dispose" -> {
                val id = call.argument<Int>("id")!!
                PlayerRegistry.remove(id)?.dispose()
                result.success(null)
            }
            "preload" -> {
                handlePreload(call, result)
            }
            "configureCache" -> {
                val maxSizeBytes = (call.argument<Number>("maxSizeBytes"))?.toLong()
                    ?: (500L * 1024 * 1024)
                VideoCache.configure(binding.applicationContext, maxSizeBytes)
                result.success(null)
            }
            "disposeAll" -> {
                PlayerRegistry.disposeAll()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    /**
     * Preloads video metadata and initial buffer data by creating a
     * temporary ExoPlayer, preparing the media, and releasing it once
     * ready. The OS-level network/disk cache retains the buffered data
     * so that a real player starts faster.
     */
    private fun handlePreload(call: MethodCall, result: MethodChannel.Result) {
        val clipsRaw = call.argument<List<Map<String, Any?>>>("clips")
        if (clipsRaw.isNullOrEmpty()) {
            result.success(null)
            return
        }

        val context: Context = binding.applicationContext
        val preloadPlayer = ExoPlayer.Builder(context)
            .setMediaSourceFactory(
                DefaultMediaSourceFactory(VideoCache.dataSourceFactory(context)),
            )
            .build()

        val mediaItems = clipsRaw.mapNotNull { map ->
            val uri = map["uri"] as? String ?: return@mapNotNull null
            val startMs = (map["startMs"] as? Number)?.toLong() ?: 0L
            val endMs = (map["endMs"] as? Number)?.toLong()

            MediaItem.Builder().setUri(uri)
                .setClippingConfiguration(
                    MediaItem.ClippingConfiguration.Builder()
                        .setStartPositionMs(startMs)
                        .apply { if (endMs != null) setEndPositionMs(endMs) }
                        .build(),
                )
                .build()
        }

        if (mediaItems.isEmpty()) {
            preloadPlayer.release()
            result.success(null)
            return
        }

        preloadPlayer.addListener(object : Player.Listener {
            override fun onPlaybackStateChanged(playbackState: Int) {
                if (playbackState == Player.STATE_READY ||
                    playbackState == Player.STATE_IDLE
                ) {
                    preloadPlayer.release()
                    result.success(null)
                }
            }

            override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
                preloadPlayer.release()
                result.success(null)
            }
        })

        preloadPlayer.setMediaItems(mediaItems)
        preloadPlayer.prepare()
    }
}

/**
 * Global registry so that [DivineVideoPlayerViewFactory] can find
 * instances created by [DivineVideoPlayerPlugin].
 */
internal object PlayerRegistry {
    private val players = ConcurrentHashMap<Int, DivineVideoPlayerInstance>()

    fun get(id: Int): DivineVideoPlayerInstance? = players[id]
    fun put(id: Int, instance: DivineVideoPlayerInstance) { players[id] = instance }
    fun remove(id: Int): DivineVideoPlayerInstance? = players.remove(id)
    fun forAll(action: (DivineVideoPlayerInstance) -> Unit) {
        players.values.toList().forEach(action)
    }
    val size: Int get() = players.size
    fun disposeAll() {
        players.values.toList().forEach { it.dispose() }
        players.clear()
    }
}
